pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +121,67 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	if pl.enableReadWriteOncePod {
+		return pl.isReadWriteOncePodAccessModeConflict(pod)
+	}
+	return framework.NewStatus(framework.Success)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable.
+// TODO(#103132): Mark pod as Unschedulable and add preemption logic.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, "error while getting node info")
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonPVCNotFound)
+			}
+			return framework.AsStatus(err)
+		}
+
+		if !v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) {
+			continue
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		key := pod.Namespace + "/" + volume.PersistentVolumeClaim.ClaimName
+		found := false
+
+		processNode := func(i int) {
+			nodeInfo := nodeInfos[i]
+			refCount := nodeInfo.PVCRefCounts[key]
+			// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+			if refCount > 0 {
+				found = true
```

Comment:
using atomic here is probably safer

Reply:
Done, using uint32 to track # of conflicts